src/rasta/c/rasta_new.c

```diff
@@ -1128,408 +1152,266 @@ void handle_retrdata(struct rasta_receive_handle *h, struct rasta_connection *co
+}
 
-                unsigned long current_time = cur_timestamp();
+char heartbeat_send_event(void * carry_data) {
+    struct timed_event_data * data = carry_data;
+    struct rasta_heartbeat_handle *h = (struct rasta_heartbeat_handle*) data->handle;
 
-                // check if T_i is expired
-                if (connection->ti_start_time > 0 && (current_time - connection->ti_start_time) >= connection->t_i){
-                    logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA HEARTBEAT", "T_i timer expired - send DisconnectionRequest");
+    struct rasta_connection * connection = rastalist_getConnection(h->connections, data->connection_index);
 
-                    // fire heartbeat timeout event
-                    fire_on_hearbeat_timeout(sr_create_notification_result(h->handle, connection));
 
-                    // T_i expired -> close connection
-                    sr_close_connection(connection,h->handle,h->mux,h->info, RASTA_DISC_REASON_TIMEOUT, 0);
-                }
+    if (connection->hb_locked) {
+        return 0;
+    }
 
-                // check if it is time to send a new heartbeat, but only if hb_stopped flag is not set
-                if(((current_time - connection->ts_last_message) >h->config.t_h) && !connection->hb_stopped){
-                    // time to send a heartbeat
-                    send_Heartbeat(h->mux,connection);
-                    // update the last message timestamp
-                    connection->ts_last_message = cur_timestamp();
-
-                    logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA HEARTBEAT", "Heartbeat sent to %d", connection->remote_id);
-                } else {
-                    sleep_time = sleep_time < (h->config.t_h - (current_time - connection->ts_last_message))*1000 ?
-                            sleep_time : (h->config.t_h - (current_time - connection->ts_last_message))*1000;
-                }
-            }
 
-            pthread_mutex_unlock(&connection->lock);
-            usleep(1000);
-        }
+    //connection is valid, check current state
+    if (connection->current_state == RASTA_CONNECTION_UP
+        || connection->current_state == RASTA_CONNECTION_RETRREQ
+        || connection->current_state == RASTA_CONNECTION_RETRRUN) {
+        send_Heartbeat(h->mux, connection, 0);
 
-        // to avoid to much CPU utilization, force context switch by sleeping for 0ns
-        //nanosleep((const struct timespec[]){{0, 0L}}, NULL);
-        usleep((useconds_t)sleep_time);
+        logger_log(h->logger, LOG_LEVEL_DEBUG, "RaSTA HEARTBEAT", "Heartbeat sent to %d", connection->remote_id);
     }
+
     return 0;
 }
 
-void * send_thread(void * handle){
-    // enable the possibility to cancel this thread
-    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
-
-    // set type to async mode, i.e the thread will be cancelled as soon as pthread_cancel is called
-    pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL);
+// TODO: split up this mess of a function
```

Comment:
Instead of leaving a `// TODO` in the (final) code, we should add a GitHub issue for this